### PR TITLE
[ML] fixing rare `_close` API timeout bug

### LINKED ISSUE #65947: "Setting upgrade_mode to enabled" yml test failing in 7.x and master
**Build scan**:

https://gradle-enterprise.elastic.co/s/7adiokzxinsqg

**Repro line**:

```
REPRODUCE WITH: ./gradlew ':x-pack:plugin:yamlRestTest' --tests "org.elasticsearch.xpack.test.rest.XPackRestIT.test {p0=ml/set_upgrade_mode/Setting upgrade_mode to enabled}" \
  -Dtests.seed=DE059F860BFE3AA5 \
  -Dtests.security.manager=true \
  -Dtests.locale=et-EE \
  -Dtests.timezone=Asia/Jerusalem \
  -Druntime.java=8 \
  -Dtests.rest.blacklist=getting_started/10_monitor_cluster_health/*

REPRODUCE WITH: ./gradlew ':x-pack:plugin:yamlRestTest' --tests "org.elasticsearch.xpack.test.rest.XPackRestIT.test {p0=ml/set_upgrade_mode/Setting upgrade_mode to enabled}" \
  -Dtests.seed=DE059F860BFE3AA5 \
  -Dtests.security.manager=true \
  -Dtests.locale=et-EE \
  -Dtests.timezone=Asia/Jerusalem \
  -Druntime.java=8 \
  -Dtests.rest.blacklist=getting_started/10_monitor_cluster_health/*
```

**Reproduces locally?**:

No

**Applicable branches**:

`7.x` `master`

**Failure history**:

Test failures that could be related to this problem started on Dec 4.

**Failure excerpt**:

Not very informative, need to look into server logs.

```
14:25:47 org.elasticsearch.xpack.test.rest.XPackRestIT > test {p0=ml/set_upgrade_mode/Setting upgrade_mode to enabled} FAILED
14:25:47     java.lang.RuntimeException: Had to resort to force-closing jobs, something went wrong?
14:25:47         at __randomizedtesting.SeedInfo.seed([DE059F860BFE3AA5:5651A05CA502575D]:0)
14:25:47         at org.elasticsearch.xpack.core.ml.integration.MlRestTestStateCleaner.deleteAllJobs(MlRestTestStateCleaner.java:86)
14:25:47         at org.elasticsearch.xpack.core.ml.integration.MlRestTestStateCleaner.clearMlMetadata(MlRestTestStateCleaner.java:31)
14:25:47         at org.elasticsearch.xpack.test.rest.AbstractXPackRestTest.clearMlState(AbstractXPackRestTest.java:226)
14:25:47         at org.elasticsearch.xpack.test.rest.AbstractXPackRestTest.cleanup(AbstractXPackRestTest.java:210)
14:25:47         at sun.reflect.GeneratedMethodAccessor16.invoke(Unknown Source)
14:25:47         at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
14:25:47         at java.lang.reflect.Method.invoke(Method.java:498)
14:25:47         at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1750)
14:25:47         at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
14:25:47         at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
14:25:47         at org.apache.lucene.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:49)
14:25:47         at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)
14:25:47         at org.apache.lucene.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:48)
14:25:47         at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
14:25:47         at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
14:25:47         at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
14:25:47         at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368)
14:25:47         at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:817)
14:25:47         at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:468)
14:25:47         at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:947)
14:25:47         at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:832)
14:25:47         at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:883)
14:25:47         at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:894)
14:25:47         at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)
14:25:47         at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
14:25:47         at org.apache.lucene.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:41)
14:25:47         at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
14:25:47         at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
14:25:47         at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
14:25:47         at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
14:25:47         at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
14:25:47         at org.apache.lucene.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
14:25:47         at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
14:25:47         at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
14:25:47         at org.apache.lucene.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:54)
14:25:47         at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
14:25:47         at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368)
14:25:47         at java.lang.Thread.run(Thread.java:748)
14:25:47 
14:25:47         Caused by:
14:25:47         java.net.SocketTimeoutException: 60 000 milliseconds timeout on connection http-outgoing-1 [ACTIVE]
14:25:47             at org.elasticsearch.client.RestClient.extractAndWrapCause(RestClient.java:864)
14:25:47             at org.elasticsearch.client.RestClient.performRequest(RestClient.java:283)
14:25:47             at org.elasticsearch.client.RestClient.performRequest(RestClient.java:270)
14:25:47             at org.elasticsearch.xpack.core.ml.integration.MlRestTestStateCleaner.deleteAllJobs(MlRestTestStateCleaner.java:78)
14:25:47             ... 37 more
14:25:47 
14:25:47             Caused by:
14:25:47             java.net.SocketTimeoutException: 60 000 milliseconds timeout on connection http-outgoing-1 [ACTIVE]
14:25:47                 at org.apache.http.nio.protocol.HttpAsyncRequestExecutor.timeout(HttpAsyncRequestExecutor.java:387)
14:25:47                 at org.apache.http.impl.nio.client.InternalIODispatch.onTimeout(InternalIODispatch.java:92)
14:25:47                 at org.apache.http.impl.nio.client.InternalIODispatch.onTimeout(InternalIODispatch.java:39)
14:25:47                 at org.apache.http.impl.nio.reactor.AbstractIODispatch.timeout(AbstractIODispatch.java:175)
14:25:47                 at org.apache.http.impl.nio.reactor.BaseIOReactor.sessionTimedOut(BaseIOReactor.java:261)
14:25:47                 at org.apache.http.impl.nio.reactor.AbstractIOReactor.timeoutCheck(AbstractIOReactor.java:502)
14:25:47                 at org.apache.http.impl.nio.reactor.BaseIOReactor.validate(BaseIOReactor.java:211)
14:25:47                 at org.apache.http.impl.nio.reactor.AbstractIOReactor.execute(AbstractIOReactor.java:280)
14:25:47                 at org.apache.http.impl.nio.reactor.BaseIOReactor.execute(BaseIOReactor.java:104)
14:25:47                 at org.apache.http.impl.nio.reactor.AbstractMultiworkerIOReactor$Worker.run(AbstractMultiworkerIOReactor.java:591)
14:25:47                 ... 1 more

```

### PR DESCRIPTION
**Pull request by benwtrent**
There is a sneaky bug that occurs in testing when we close jobs.

It appears that if a job is just recently assigned to a node (and in the process of reopening). When this occurs a call to `_close` will eventually timeout.

We don’t actually close the job if it is assigned to a node but does not yet have a native process. There is a window where we ignore the `CLOSING` state and the continue to create the process context and even overwrite the `CLOSING` state back to OPENED.

The close API thinks the job is already closed because it doesn’t have a native process. The job continues executing as it never checks again if it is closing. The model snapshot revert that was added to when a job starts executing on a node has only exacerbated an existing bug.

End users could potentially see this bug, but the recovery from it is easy. They can simply call `_close` again and it will work. 

closes #65947

### CODE REVIEW

**Comment by benwtrent** [comment]
run elasticsearch-ci/2

**Comment by benwtrent** [comment]
I also think this PR fixes the test failures here: https://github.com/elastic/elasticsearch/issues/63980

The job `opening` after a `_close` occurs is EXACTLY what this bug is. But I will hold off unmuting that test until after this PR proves to correct the issue for the upgrade mode tests.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/AutodetectProcessManager.java [hunk 1]
```diff
@@ -521,6 +521,12 @@ protected void doRun() {
                         logger.debug("Aborted opening job [{}] as it has been closed", job.getId());
                         return;
                     }
+                    // We check again after the process state is locked, but doing a which `isClosing` check here
```
Comment (dimitris-athanasiou) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/AutodetectProcessManager.java:524]: This comment is a bit confusing. Could you revisit it please?

**Review by dimitris-athanasiou — APPROVED** [review]
LGTM I just left a comment about improving a comment that reads hard.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/AutodetectProcessManager.java (hunk 1 above)
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/AutodetectProcessManager.java:524]:
```suggestion
                    // We check again after the process state is locked to ensure no race conditions are hit.
```

**Pull request merged**